src/env.rs

```diff
@@ -0,0 +1,296 @@
+///
+/// All widget methods have access to an environment, and it is passed
+/// downwards during traversals.
+///
+/// A widget can retrieve theme parameters (colors, dimensions, etc.). In
+/// addition, it can pass custom data down to all descendants. An important
+/// example of the latter is setting a value for enabled/disabled status
+/// so that an entire subtree can be disabled ("grayed out") with one
+/// setting.
+#[derive(Clone, Default)]
+pub struct Env(Arc<EnvImpl>);
+
+#[derive(Clone, Default)]
+struct EnvImpl {
+    map: HashMap<String, Value>,
+}
+
+/// A typed key.
+///
+/// This lets you retrieve values of a given type. The parameter
+/// implements [`ValueType`]. For "expensive" types, this is a reference,
+/// so the type for a string is `Key<&str>`.
+///
+/// [`ValueType`]: trait.ValueType.html
+pub struct Key<T> {
+    key: &'static str,
+    value_type: PhantomData<T>,
+}
+
+// we could do some serious deriving here: the set of types that can be stored
+// could be defined per-app
+// Also consider Box<Any> (though this would also impact debug).
+/// A dynamic type representing all values that can be stored in an environment.
+#[derive(Clone)]
+pub enum Value {
+    Point(Point),
+    Size(Size),
+    Rect(Rect),
+    Color(Color),
+    Float(f64),
+    String(String),
+}
+
+/// Values which can be stored in an environment.
+///
+/// Note that for "expensive" types this is the reference. For example,
+/// for strings, this trait is implemented on `&'a str`. The trait is
+/// parametrized on a lifetime so that it can be used for references in
+/// this way.
+pub trait ValueType<'a>: Sized {
+    /// The corresponding owned type.
+    type Owned;
+
+    /// Attempt to convert the generic `Value` into this type.
+    fn try_from_value(v: &'a Value) -> Result<Self, String>;
+}
+
+impl Env {
+    /// Gets a value from the environment, expecting it to be present.
+    ///
+    /// Note that the return value is a reference for "expensive" types such
+    /// as strings, but an ordinary value for "cheap" types such as numbers
+    /// and colors.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the key is not found, or if it is present with the wrong type.
+    pub fn get<'a, V: ValueType<'a>>(&'a self, key: Key<V>) -> V {
+        if let Some(value) = self.0.map.get(key.key) {
+            value.to_inner_unchecked()
+        } else {
+            panic!("key for {} not found", key.key)
+        }
+    }
+
+    /// Gets a value from the environment.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the value for the key is found, but has the wrong type.
+    pub fn try_get<'a, V: ValueType<'a>>(&'a self, key: Key<V>) -> Option<V> {
+        self.0
+            .map
+            .get(key.key)
+            .map(|value| value.to_inner_unchecked())
+    }
+
+    /// Adds a key/value, acting like a builder.
+    pub fn adding<'a, V: ValueType<'a>>(mut self, key: Key<V>, value: impl Into<V::Owned>) -> Env
```

Comment:
I'm not sure I love mixing generic syntax with argument position `impl`? It might be clearer here just as,

```rust
pub fn adding<'a, V, T>(mut self, key: Key<V>, value: T) -> Env
where
    V: ValueType<'a>,
    V::Owned: Into<Value>,
    T: Into<V::Owned>,
{
// ...
}
```

But I don't feel strongly about this.

This also makes me wonder if we couldn't have the `Into<Value>` bound be a bound on the associated type in the trait? This would clean up these function signatures somewhat.

Reply:
Yeah, I think `Into<Value>` on `ValueType::Owned` is a win. Here's the minor edit:

```rust
    pub fn adding<'a, V: ValueType<'a>>(mut self, key: Key<V>, value: impl Into<V::Owned>) -> Env
```

And here's what it would be with no impl but on one line:

```rust
pub fn adding<'a, V: ValueType<'a>, T: Into<V::Owned>>(mut self, key: Key<V>, value: T) -> Env
```

And with 'where':

```rust
    pub fn adding<'a, V, T>(mut self, key: Key<V>, value: T) -> Env
    where
        V: ValueType<'a>,
        T: Into<V::Owned>,
```

Preference?

Reply:
The first or the last, first is cleaner but depends how we feel about mixing `impl` args with normal generics. Your call.

Reply:
I'm gonna go with the first, because `impl Into` feels really idiomatic to me. The third is also appealing because it reads a little less like a "line noise" type.